File: packages/firestore/ios/Plugin/Classes/Results/GetDocumentResult.swift
import Foundation
import FirebaseFirestore
import Capacitor

@objc public class GetDocumentResult: NSObject, Result {
    let documentSnapshot: DocumentSnapshot
    let serverTimestampBehavior: ServerTimestampBehavior

    init(_ documentSnapshot: DocumentSnapshot, _ serverTimestampBehavior: ServerTimestampBehavior = .none) {
        self.documentSnapshot = documentSnapshot
        self.serverTimestampBehavior = serverTimestampBehavior
    }

    public func toJSObject() -> AnyObject {
        let snapshotDataResult = FirebaseFirestoreHelper.createJSObjectFromHashMap(documentSnapshot.data(with: serverTimestampBehavior))

        var snapshotResult = JSObject()
        snapshotResult["id"] = documentSnapshot.documentID
        snapshotResult["path"] = documentSnapshot.reference.path
        if let snapshotDataResult = snapshotDataResult {
            snapshotResult["data"] = snapshotDataResult
        } else {
            snapshotResult["data"] = NSNull()
        }

        var metadata = JSObject()
        metadata["fromCache"] = documentSnapshot.metadata.isFromCache
        metadata["hasPendingWrites"] = documentSnapshot.metadata.hasPendingWrites
        snapshotResult["metadata"] = metadata

        var result = JSObject()
        result["snapshot"] = snapshotResult
        return result as AnyObject
    }
}

File: packages/firestore/android/src/main/java/app/capgo/capacitor/firebase/firestore/GetDocumentResult.java
package app.capgo.capacitor.firebase.firestore.classes.results;

import com.getcapacitor.JSObject;
import com.google.firebase.firestore.DocumentSnapshot;
import app.capgo.capacitor.firebase.firestore.FirebaseFirestoreHelper;
import app.capgo.capacitor.firebase.firestore.interfaces.Result;
import org.json.JSONObject;

public class GetDocumentResult implements Result {

    private DocumentSnapshot documentSnapshot;

    public GetDocumentResult(DocumentSnapshot documentSnapshot) {
        this.documentSnapshot = documentSnapshot;
    }

    public JSObject toJSObject() {
        Object snapshotDataResult = FirebaseFirestoreHelper.createJSObjectFromMap(documentSnapshot.getData());

        JSObject snapshotResult = new JSObject();
        snapshotResult.put("id", documentSnapshot.getId());
        snapshotResult.put("path", documentSnapshot.getReference().getPath());
        if (snapshotDataResult == null) {
            snapshotResult.put("data", JSONObject.NULL);
        } else {
            snapshotResult.put("data", snapshotDataResult);
        }
        snapshotResult.put("metadata", FirebaseFirestoreHelper.createSnapshotMetadataResult(documentSnapshot));

        JSObject result = new JSObject();
        result.put("snapshot", snapshotResult);
        return result;
    }
}
